Source/ARTRealtimeChannel.m

```diff
@@ -331,23 +323,79 @@ - (void)publishProtocolMessage:(ARTProtocolMessage *)pm callback:(void (^)(ARTSt
             if (cb) {
-                cb(statusInvalidChannel);
+                cb(invalidChannelError);
             }
             break;
         }
-        case ARTRealtimeChannelAttached:
-        {
+        case ARTRealtimeChannelAttached: {
             if (_realtime.connection.state_nosync == ARTRealtimeConnected) {
-                [self sendMessage:pm callback:cb];
-            } else {
+                [self sendMessage:pm callback:^(ARTStatus *status) {
+                    if (cb) cb(status.errorInfo);
+                }];
+            }
+            else {
                 [self addToQueue:pm callback:queuedCallback];
+            }
+            break;
+        }
+    }
+} ART_TRY_OR_MOVE_TO_FAILED_END
+}
 
-                [self.realtime.internalEventEmitter once:[ARTEvent newWithConnectionEvent:ARTRealtimeConnectionEventConnected] callback:^(ARTConnectionStateChange *__art_nullable change) {
-                    [weakSelf sendQueuedMessages];
-                }];
+- (void)publishProtocolMessage:(ARTProtocolMessage *)pm callback:(void (^)(ARTStatus *))cb {
+ART_TRY_OR_MOVE_TO_FAILED_START(_realtime) {
+    ARTStatus *statusInvalidConnectionState = [ARTStatus state:ARTStateError info:[ARTErrorInfo createWithCode:90001 message:@"channel operation failed (invalid connection state)"]];
+    ARTStatus *statusInvalidChannelState = [ARTStatus state:ARTStateError info:[ARTErrorInfo createWithCode:90001 message:@"channel operation failed (invalid channel state)"]];
+
+    __weak __typeof(self) weakSelf = self;
+    void (^queuedCallback)(ARTStatus *) = ^(ARTStatus *status) {
+        switch (weakSelf.state_nosync) {
+            case ARTRealtimeChannelSuspended:
+            case ARTRealtimeChannelFailed:
+                if (cb) {
+                    cb(status.state == ARTStateOk ? statusInvalidChannelState : status);
+                }
+                return;
+            default:
+                break;
+        }
+        if (cb) {
+            cb(status);
+        }
+    };
+
+    switch (_realtime.connection.state_nosync) {
+        case ARTRealtimeClosing:
+        case ARTRealtimeClosed: {
+            if (cb) {
+                cb(statusInvalidConnectionState);
+            }
+            return;
+        }
+        default:
+            break;
+    }
+
+    switch (self.state_nosync) {
+        case ARTRealtimeChannelSuspended:
+        case ARTRealtimeChannelFailed: {
+            if (cb) {
+                cb(statusInvalidChannelState);
+            }
+            break;
+        }
+        case ARTRealtimeChannelInitialized:
+        case ARTRealtimeChannelDetaching:
+        case ARTRealtimeChannelDetached:
+        case ARTRealtimeChannelAttaching:
+        case ARTRealtimeChannelAttached: {
+            if (_realtime.connection.state_nosync == ARTRealtimeConnected) {
+                [self sendMessage:pm callback:cb];
+            }
+            else {
+                [self addToQueue:pm callback:queuedCallback];
```

Comment:
I still don't understand why you're queueing Messages in a channel-level queue here. What's the reasoning behind doing that rather than just sending them to the connection queue? (cf previous comment https://github.com/ably/ably-cocoa/pull/795#discussion_r278043050 )